homeassistant/components/yale_smart_alarm/lock.py

```diff
@@ -0,0 +1,145 @@
+"""Lock for Yale Alarm."""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from yalesmartalarmclient.exceptions import AuthenticationError, UnknownError
+
+from homeassistant.components.lock import LockEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_CODE, CONF_CODE, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    CONF_LOCK_CODE_DIGITS,
+    COORDINATOR,
+    DEFAULT_LOCK_CODE_DIGITS,
+    DOMAIN,
+    LOGGER,
+    MANUFACTURER,
+    MODEL,
+)
+from .coordinator import YaleDataUpdateCoordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Yale lock entry."""
+
+    coordinator: YaleDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        COORDINATOR
+    ]
+    code_format = entry.options.get(CONF_LOCK_CODE_DIGITS, DEFAULT_LOCK_CODE_DIGITS)
+
+    async_add_entities(
+        YaleDoorlock(coordinator, data, code_format)
+        for data in coordinator.data["locks"]
+    )
+
+
+class YaleDoorlock(CoordinatorEntity, LockEntity):
+    """Representation of a Yale doorlock."""
+
+    def __init__(
+        self, coordinator: YaleDataUpdateCoordinator, data: dict, code_format: int
+    ) -> None:
+        """Initialize the Yale Lock Device."""
+        super().__init__(coordinator)
+        self._coordinator = coordinator
+        self._attr_name = data["name"]
+        self._attr_unique_id = data["address"]
+        self._attr_device_info = DeviceInfo(
+            name=self._attr_name,
+            manufacturer=MANUFACTURER,
+            model=MODEL,
+            identifiers={(DOMAIN, data["address"])},
+            via_device=(DOMAIN, self._coordinator.entry.data[CONF_USERNAME]),
+        )
+        self._attr_code_format = f"^\\d{code_format}$"
+
+    async def async_unlock(self, **kwargs) -> None:
+        """Send unlock command."""
+        if TYPE_CHECKING:
```

Comment:
That's much better.  

You could drop the if type checking since there is no circular import here and it would be a legitimate implementation error if it was missing at run time that would be good to report.

Reply:
It's only there to make mypy happy (as if yale is not available it will already raise an error from the coordinator) and a recommendation to use this syntax instead from Martin in the #63672 PR

Reply:
My preference is to not use if TYPE_CHECKING unless it's strictly needed since I want the code to always tell me when there has been an implementation error even if it's at runtime. 

It's not something that should block this though

Reply:
Thanks. I intend to clean this up further in upcoming PR.